### PR TITLE
Support caller-allocated buffer for stateless collection marshalling

### PR DESCRIPTION
**Pull request by elinor-fung**
- Refactor blittable element handling into a base helper for shared functionality between stateless and stateful
- Enable using a caller-allocated buffer for stateless collection marshalling
  - Shape data was already determined, so this is just adds the marshalling strategy/generation side

### CODE REVIEW

**Review by jkoritzinsky — APPROVED** [review]

**Review by jkoritzinsky — APPROVED** [review]

File: src/libraries/System.Runtime.InteropServices/tests/TestAssets/SharedTypes/NonBlittable.cs [hunk 1]
```diff
@@ -422,6 +412,37 @@ public static void Free(byte* unmanaged)
...
+            if (managed is null)
+            {
+                numElements = 0;
+                return null;
+            }
+
+            numElements = managed.Count;
+
+            // Always allocate at least one byte when the list is zero-length.
+            int spaceRequired = Math.Max(checked(sizeof(TUnmanagedElement) * numElements), 1);
```
Comment (AaronRobinsonMSFT) [inline src/libraries/System.Runtime.InteropServices/tests/TestAssets/SharedTypes/NonBlittable.cs:432]: Do we need this at all? I'm confused by the comment.

File: src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/Marshalling/StatelessMarshallingStrategy.cs [hunk 2]
```diff
@@ -277,92 +301,40 @@ public TypeSyntax AsNativeType(TypePositionInfo info)
...
-                            _marshallerTypeSyntax,
-                            IdentifierName(ShapeMemberNames.LinearCollection.Stateless.AllocateContainerForUnmanagedElements)),
-                        ArgumentList(SeparatedList(new ArgumentSyntax[]
-                        {
-                            Argument(IdentifierName(managedIdentifier)),
-                            Argument(IdentifierName(numElementsIdentifier))
-                                .WithRefOrOutKeyword(Token(SyntaxKind.OutKeyword))
-                        })))));
+            if (_shape.HasFlag(MarshallerShape.ToUnmanaged)
+                && !(_shape.HasFlag(MarshallerShape.CallerAllocatedBuffer) && MarshallerHelpers.CanUseCallerAllocatedBuffer(info, context)))
```
Comment (AaronRobinsonMSFT) [inline src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/Marshalling/StatelessMarshallingStrategy.cs:308]:
```suggestion
                && !(_shape.HasFlag(MarshallerShape.CallerAllocatedBuffer)
                    && MarshallerHelpers.CanUseCallerAllocatedBuffer(info, context)))
```

**Review by AaronRobinsonMSFT — APPROVED** [review]

File: src/libraries/System.Runtime.InteropServices/tests/TestAssets/SharedTypes/NonBlittable.cs (hunk 1 above)
Comment (elinor-fung) [reply src/libraries/System.Runtime.InteropServices/tests/TestAssets/SharedTypes/NonBlittable.cs:432]: Bad copy-pasting..

Comment (elinor-fung) [inline src/libraries/System.Runtime.InteropServices/tests/TestAssets/SharedTypes/NonBlittable.cs:432]:
```suggestion
            int spaceRequired = checked(sizeof(TUnmanagedElement) * numElements);
```

**Pull request merged**